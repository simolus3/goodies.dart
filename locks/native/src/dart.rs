use std::{ffi::CStr, marker::PhantomData};

use dart_sys::{
    Dart_CObject, Dart_CObject_Type_Dart_CObject_kArray, Dart_CObject_Type_Dart_CObject_kBool,
    Dart_CObject_Type_Dart_CObject_kNull, Dart_CObject_Type_Dart_CObject_kString, Dart_Port_DL,
    Dart_PostCObject_DL,
};

#[repr(transparent)]
pub struct DartPort(Dart_Port_DL);

impl From<Dart_Port_DL> for DartPort {
    fn from(value: Dart_Port_DL) -> Self {
        return Self(value);
    }
}

impl DartPort {
    pub fn send<'a>(&self, message: &mut DartObject<'a>) -> bool {
        let raw = &mut message.raw;
        let raw = std::ptr::from_mut(raw);

        unsafe { Dart_PostCObject_DL.unwrap_unchecked()(self.0, raw) }
    }
}

#[repr(transparent)]
pub struct DartObject<'a> {
    raw: Dart_CObject,
    references: PhantomData<&'a ()>,
}

impl<'a> DartObject<'a> {
    pub const NULL: DartObject<'static> = DartObject {
        raw: Dart_CObject {
            type_: Dart_CObject_Type_Dart_CObject_kNull,
            value: dart_sys::_Dart_CObject__bindgen_ty_1 { as_bool: false },
        },
        references: PhantomData,
    };

    pub fn array(elements: &'a mut [&'a DartObject<'a>]) -> Self {
        Self {
            raw: Dart_CObject {
                type_: Dart_CObject_Type_Dart_CObject_kArray,
                value: dart_sys::_Dart_CObject__bindgen_ty_1 {
                    as_array: dart_sys::_Dart_CObject__bindgen_ty_1__bindgen_ty_3 {
                        length: elements.len() as isize,
                        // This cast is safe because references are pointers at runtime, and since
                        // DartObject is a repr(transparent) wrapper around Dart_CObject.
                        values: elements.as_mut_ptr().cast(),
                    },
                },
            },
            references: PhantomData,
        }
    }
}

impl<'a> From<&'a CStr> for DartObject<'a> {
    fn from(value: &'a CStr) -> Self {
        Self {
            raw: Dart_CObject {
                type_: Dart_CObject_Type_Dart_CObject_kString,
                value: dart_sys::_Dart_CObject__bindgen_ty_1 {
                    as_string: value.as_ptr(),
                },
            },
            references: PhantomData,
        }
    }
}

impl From<bool> for DartObject<'static> {
    fn from(value: bool) -> Self {
        Self {
            raw: Dart_CObject {
                type_: Dart_CObject_Type_Dart_CObject_kBool,
                value: dart_sys::_Dart_CObject__bindgen_ty_1 { as_bool: value },
            },
            references: PhantomData,
        }
    }
}
