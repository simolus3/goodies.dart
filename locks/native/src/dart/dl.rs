use std::{
    ffi::{CStr, c_char, c_int},
    os::raw::c_void,
};

use crate::dart::object::RawDartCObject;

#[repr(C)]
struct RawDartApiEntry {
    pub name: *const c_char,
    pub function: *const c_void,
}

#[repr(C)]
struct RawDartApi {
    pub major: c_int,
    pub minor: c_int,
    pub functions: *const RawDartApiEntry,
}

pub struct DartApi {
    pub post_object: unsafe extern "C" fn(port_id: i64, message: *mut RawDartCObject) -> bool,
}

impl DartApi {
    pub unsafe fn from_raw(ptr: *const c_void) -> Self {
        let raw = unsafe { (ptr.cast::<RawDartApi>()).as_ref() }.unwrap();

        if raw.major != 2 || raw.minor < 6 {
            panic!("Incompatible Dart embedder, need version ^2.6");
        }

        let mut post_object: Option<PostObjectSignature> = None;

        // See https://github.com/dart-lang/sdk/blob/9f2218e4509210576abecc79d7184c94a3ad0e26/runtime/include/dart_api_dl.c#L23-L30
        let mut entry = raw.functions;
        loop {
            {
                let Some(entry) = (unsafe { entry.as_ref() }) else {
                    break;
                };
                if entry.name.is_null() {
                    break;
                }

                let name = unsafe { CStr::from_ptr(entry.name) };
                let function = entry.function;

                if name == c"Dart_PostCObject" {
                    post_object = unsafe { function.cast::<Option<PostObjectSignature>>().read() };
                }
            }

            entry = unsafe { entry.add(1) };
        }

        Self {
            post_object: post_object.unwrap(),
        }
    }
}

type PostObjectSignature = unsafe extern "C" fn(port_id: i64, message: *mut RawDartCObject) -> bool;
