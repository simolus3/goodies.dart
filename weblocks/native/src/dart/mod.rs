mod dl;
mod object;

pub use dl::DartApi;
pub use object::DartObject;

/// A wrapper around a native `SendPort`.
#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(transparent)]
pub struct DartPort(i64);

impl DartPort {
    /// Sends a `message` to this port.
    ///
    /// Returns true if the external contents in `message` have been moved to Dart, false otherwise.
    pub fn send<'a>(&self, api: &DartApi, message: &mut DartObject<'a>) -> bool {
        let raw = &mut message.raw;
        let raw = std::ptr::from_mut(raw);

        unsafe { (api.post_object)(self.0, raw) }
    }
}
