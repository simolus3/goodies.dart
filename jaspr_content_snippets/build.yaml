builders:
  api_index:
    import: 'package:jaspr_content_snippets/builders.dart'
    build_to: cache
    builder_factories: ['DartIndexBuilder.new']
    auto_apply: all_packages
    runs_before: [jaspr_content_snippets]
    build_extensions: {"lib/lib": ['api.json']}

  jaspr_content_snippets:
    import: 'package:jaspr_content_snippets/builders.dart'
    build_to: cache
    builder_factories: ['excerptsBuilder']
    auto_apply: none
    build_extensions: {'': ['.snippet.json']}
    defaults:
      options:
        process_without_directives: false