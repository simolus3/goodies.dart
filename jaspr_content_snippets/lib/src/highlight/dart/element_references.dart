import 'package:analyzer/dart/ast/syntactic_entity.dart';
import 'package:analyzer/dart/element/element.dart';
import 'package:build/build.dart';

import '../../dartdoc/dart_index.dart';
import '../../dartdoc/uris.dart';
import '../highlighter.dart';

/// For identifier tokens fond by the highlights computer, asynchronously
/// resolves suitable `dart doc` links.
final class ElementReferences {
  final _pending = <_PendingElementReference>[];

  void trackElement(SyntacticEntity token, Element element) {
    _pending.add(_PendingElementReference(token.offset, token.length, element));
  }

  Future<void> resolveAgainstSortedTokens(
    List<HighlightToken> tokens,
    BuildStep buildStep,
  ) async {
    final index = await DartIndex.of(buildStep);

    for (final reference in _findReferences(tokens)) {
      final import = await index.publicLibraryForElement(
        reference.element,
        buildStep,
      );

      if (import != null) {
        reference.token.documentationUri = _dartDocUri(
          import,
          reference.element,
        );
      }
    }
  }

  Iterable<_MatchedReference> _findReferences(
    List<HighlightToken> tokens,
  ) sync* {
    // Sort references with the same comparator as tokens.
    _pending.sort((a, b) {
      // First sort by offset.
      if (a.offset != b.offset) {
        return a.offset.compareTo(b.offset);
      }

      // Then length (so longest are first).
      if (a.length != b.length) {
        return -a.length.compareTo(b.length);
      }

      // Apart from these we don't need a stable sort here.
      return 0;
    });

    var iterator = _pending.iterator;
    if (!iterator.moveNext()) {
      return;
    }

    for (final token in tokens) {
      // Find an element reference matching this token.
      while (iterator.current.offset < token.offset) {
        if (!iterator.moveNext()) {
          return;
        }
      }

      if (iterator.current.offset == token.offset &&
          iterator.current.length == token.length) {
        yield _MatchedReference(iterator.current.element, token);
      }
    }
  }

  Uri _dartDocUri(PublicLibrary library, Element element) {
    // TODO: Support overriding dartdoc base URIs?
    final base = defaultDocumentationUri(library.id.package);

    return documentationForElement(element, library.dirName, base);
  }
}

final class _PendingElementReference {
  final int offset;
  final int length;
  final Element element;

  _PendingElementReference(this.offset, this.length, this.element);
}

final class _MatchedReference {
  final Element element;
  final HighlightToken token;

  _MatchedReference(this.element, this.token);
}
