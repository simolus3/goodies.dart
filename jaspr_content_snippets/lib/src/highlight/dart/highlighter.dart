import 'package:analyzer/dart/analysis/results.dart';
import 'package:build/build.dart';

import '../highlighter.dart';
import 'computer_highlights.dart';

final class DartHighlighter implements Highlighter {
  final BuildStep buildStep;

  DartHighlighter(this.buildStep);

  @override
  Future<List<HighlightToken>> highlight(AssetId id) async {
    final library = await buildStep.resolver.libraryFor(
      id,
      allowSyntaxErrors: true,
    );
    final path = library.session.uriConverter.uriToPath(library.uri)!;
    final resolveResult = switch (await library.session.getResolvedUnit(path)) {
      final ResolvedUnitResult result => result,
      final other => throw StateError('Could not resolve $path: $other'),
    };

    final computer = DartUnitHighlightsComputer(resolveResult.unit);
    final semanticTokens = computer.computeSemanticTokens();
    semanticTokens.sort(HighlightToken.offsetLengthPrioritySort);

    // Add dartdoc references to found tokens
    await computer.referencedElements.resolveAgainstSortedTokens(
      semanticTokens,
      buildStep,
    );

    return semanticTokens;
  }
}
