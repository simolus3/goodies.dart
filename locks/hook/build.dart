import 'package:code_assets/code_assets.dart';
import 'package:hooks/hooks.dart';

void main(List<String> args) async {
  await build(args, (input, output) async {
    input.config.code.targetOS;
    output.assets.code.add(
      CodeAsset(
        package: 'locks',
        name: 'src/native/bindings.dart',
        linkMode: DynamicLoadingSystem(
          Uri.file(
            '/Users/simon/src/goodies.dart/locks/native/target/debug/libdart_locks.dylib',
          ),
        ),
      ),
    );
  });
}
