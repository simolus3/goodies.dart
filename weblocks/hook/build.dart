import 'dart:io';

import 'package:code_assets/code_assets.dart';
import 'package:hooks/hooks.dart';
import 'package:path/path.dart' as p;

void main(List<String> args) async {
  await build(args, (input, output) async {
    Future<void> useHostBuild() async {
      await _cargoBuild();

      final name = switch (Platform.operatingSystem) {
        'linux' => 'libdart_locks.so',
        'macos' => 'libdart_locks.dylib',
        'windows' => 'dart_locks.dll',
        _ => throw 'Unknown operating system',
      };
      final path = p.absolute('native/target/debug/$name');

      output.assets.code.add(
        CodeAsset(
          package: 'weblocks',
          name: 'src/native/bindings.dart',
          linkMode: DynamicLoadingSystem(Uri.file(path)),
        ),
      );

      final native = Directory('native');
      final dependencies = native.listSync().expand<String>((entry) sync* {
        if (p.basename(entry.path) == 'target') {
          return;
        }

        if (entry is Directory) {
          yield* entry
              .listSync(recursive: true)
              .whereType<File>()
              .map((e) => e.path);
        } else if (entry is File) {
          yield entry.path;
        }
      });
      output.dependencies.addAll(dependencies.map((p) => Uri.file(p)));
    }

    final localBuild = input.userDefines['local_build'] == true;
    if (localBuild || !Directory('assets').existsSync()) {
      await useHostBuild();
    } else {
      final os = input.config.code.targetOS;
      final builds = switch (os) {
        OS.macOS => _macosBuilds,
        OS.iOS =>
          input.config.code.iOS.targetSdk == IOSSdk.iPhoneOS
              ? _iosBuilds
              : _iosSimulatorBuilds,
        OS.android => _androidBuilds,
        OS.linux => _linuxBuilds,
        OS.windows => _windowsBuilds,
        OS(:final name) => throw UnsupportedError(
          'Operating system not supported: $name',
        ),
      };

      final targetArchitecture = input.config.code.targetArchitecture;
      final filename =
          builds[targetArchitecture] ??
          (throw UnsupportedError(
            'Architecture $targetArchitecture not supported for $os. Supported are: ${builds.keys}',
          ));

      output.assets.code.add(
        CodeAsset(
          package: 'weblocks',
          name: 'src/native/bindings.dart',
          file: Uri.file(p.absolute(p.join('assets', filename))),
          linkMode: DynamicLoadingBundled(),
        ),
      );
    }
  });
}

Future<void> _cargoBuild() async {
  final process = await Process.start(
    'cargo',
    ['build'],
    mode: ProcessStartMode.inheritStdio,
    workingDirectory: 'native',
  );

  final exitCode = await process.exitCode;
  if (exitCode != 0) {
    throw 'Could not invoke cargo build';
  }
}

const _androidBuilds = {
  Architecture.arm: 'libdart_locks.android_v7a.so',
  Architecture.arm64: 'libdart_locks.android_v8a.so',
  Architecture.x64: 'libdart_locks.android_x86_64.so',
};

const _iosBuilds = {Architecture.arm64: 'libdart_locks.ios_aarch64.dylib'};

const _iosSimulatorBuilds = {
  Architecture.arm64: 'libdart_locks.ios_sim_aarch64.dylib',
  Architecture.x64: 'libdart_locks.ios_sim_x64.dylib',
};

const _macosBuilds = {
  Architecture.arm64: 'libdart_locks.macos_aarch64.dylib',
  Architecture.x64: 'libdart_locks.macos_x64.dylib',
};

const _linuxBuilds = {
  Architecture.arm: 'libdart_locks.linux_arm7.so',
  Architecture.arm64: 'libdart_locks.linux_aarch64.so',
  Architecture.x64: 'libdart_locks.linux_x64.so',
  Architecture.riscv64: 'libdart_locks.linux_riscv.so',
};

const _windowsBuilds = {
  Architecture.arm64: 'dart_locks.win_aarch64.dll',
  Architecture.x64: 'dart_locks.win_x64.dll',
};
