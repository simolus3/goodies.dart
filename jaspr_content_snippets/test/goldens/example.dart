import 'dart:async';

import 'package:build/build.dart';

// #docregion outline
final class ExampleBuilder extends Builder {
  // #enddocregion outline
  // #docregion buildExtensions
  @override
  Map<String, List<String>> get buildExtensions => const {
    '.dart': ['.snippets'],
  };
  // #enddocregion buildExtensions

  // #docregion build
  @override
  FutureOr<void> build(BuildStep buildStep) {
    throw UnimplementedError();
  }

  // #enddocregion build
  // #docregion outline
}

// #enddocregion outline
