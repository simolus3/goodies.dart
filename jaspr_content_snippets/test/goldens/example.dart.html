<!DOCTYPE html>
<html>
  <head>
    <base href="/"/>
    <meta charset="utf-8"/>
    <style>
      .keyword {
        color: blue;
      }
    </style>
    <style>
      .type {
        color: red;
      }
    </style>
    <style>
      .class {
        color: darkRed;
      }
    </style>
    <style>
      .property {
        color: darkGreen;
      }
    </style>
    <style>
      .function {
        color: yellowGreen;
      }
    </style>
    <style>
      .method {
        color: yellowGreen;
      }
    </style>
    <style>
      .variable {
        color: green;
      }
    </style>
    <style>
      .parameter {
        color: green;
      }
    </style>
    <style>
      .string {
        color: cyan;
      }
    </style>
    <style>
      .number {
        color: darkCyan;
      }
    </style>
    <style>
      .annotation {
        color: lightGreen;
      }
    </style>
    <style>
      .declaration {
        font-weight: bold;
      }
    </style>
    <!--$-->
    <title>test/goldens/example.dart</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <!--/-->
  </head>
  <body>
    <div>
      <h2>(full)</h2>
      <code>
        <pre><span><span class="keyword">import</span> <span class="string">'dart:async'</span>;

<span class="keyword">import</span> <span class="string"><a href="https://pub.dev/documentation/build/latest/build/build-library.html">'package:build/build.dart'</a></span>;

<span class="keyword">final</span> <span class="keyword">class</span> <span class="class declaration">ExampleBuilder</span> <span class="keyword">extends</span> <span class="class"><a href="https://pub.dev/documentation/build/latest/build/Builder-class.html">Builder</a></span> {
  <span class="annotation">@</span><span class="property annotation">override</span>
  <span class="class">Map</span>&lt;<span class="class">String</span>, <span class="class">List</span>&lt;<span class="class">String</span>&gt;&gt; <span class="keyword">get</span> <span class="property declaration instance">buildExtensions</span> =&gt; <span class="keyword">const</span> {
    <span class="string">'.dart'</span>: [<span class="string">'.snippets'</span>],
  };

  <span class="annotation">@</span><span class="property annotation">override</span>
  <span class="class">FutureOr</span>&lt;<span class="keyword void">void</span>&gt; <span class="method declaration instance">build</span>(<span class="class"><a href="https://pub.dev/documentation/build/latest/build/BuildStep-class.html">BuildStep</a></span> <span class="parameter declaration">buildStep</span>) {
    <span class="keyword control">throw</span> <span class="class constructor">UnimplementedError</span>();
  }

}
</span></pre>
      </code>
      <hr/>
    </div>
    <div>
      <h2>outline</h2>
      <code>
        <pre><span><span class="keyword">final</span> <span class="keyword">class</span> <span class="class declaration">ExampleBuilder</span> <span class="keyword">extends</span> <span class="class"><a href="https://pub.dev/documentation/build/latest/build/Builder-class.html">Builder</a></span> {
}
</span></pre>
      </code>
      <hr/>
    </div>
    <div>
      <h2>buildExtensions</h2>
      <code>
        <pre><span>  <span class="annotation">@</span><span class="property annotation">override</span>
  <span class="class">Map</span>&lt;<span class="class">String</span>, <span class="class">List</span>&lt;<span class="class">String</span>&gt;&gt; <span class="keyword">get</span> <span class="property declaration instance">buildExtensions</span> =&gt; <span class="keyword">const</span> {
    <span class="string">'.dart'</span>: [<span class="string">'.snippets'</span>],
  };</span></pre>
      </code>
      <hr/>
    </div>
    <div>
      <h2>build</h2>
      <code>
        <pre><span>  <span class="annotation">@</span><span class="property annotation">override</span>
  <span class="class">FutureOr</span>&lt;<span class="keyword void">void</span>&gt; <span class="method declaration instance">build</span>(<span class="class"><a href="https://pub.dev/documentation/build/latest/build/BuildStep-class.html">BuildStep</a></span> <span class="parameter declaration">buildStep</span>) {
    <span class="keyword control">throw</span> <span class="class constructor">UnimplementedError</span>();
  }
</span></pre>
      </code>
      <hr/>
    </div>
  </body>
</html>
